hipscat/dask_utils.py

```diff
@@ -155,143 +155,192 @@ def _to_hips(df, hipsPath, base_filename):
+                tocull=tocull1
+            )
+
+            c2_df = util.frame_cull(
+                df=c2_df, df_md=c2_md,
+                order=order, pix=pix,
+                cols=c2_cols,
+                tocull=tocull2
+            )
+            
+            #Sometimes the c1_df or c2_df contain zero sources 
+            # after culling
+            if len(c1_df) and len(c2_df):
+
+                #calculate the xy gnomonic positions from the 
+                # pixel's center for each dataframe
+                xy1 = util.frame_gnomonic(c1_df, c1_md, clon, clat)
+                xy2 = util.frame_gnomonic(c2_df, c2_md, clon, clat)
+
+                #construct the KDTree from the comparative catalog: c2/xy2
+                tree = KDTree(xy2, leaf_size=2)
+                #find the indicies for the nearest neighbors 
+                #this is the cross-match calculation
+                dists, inds = tree.query(xy1, k=n_neighbors)
+
+                #numpy indice magic for the joining of the two catalogs
+                outIdx = np.arange(len(c1_df)*n_neighbors) # index of each row in the output table (0... number of output rows)
+                leftIdx = outIdx // n_neighbors            # index of the corresponding row in the left table (0, 0, 0, 1, 1, 1, 2, 2, 2, ...)
+                rightIdx = inds.ravel()                    # index of the corresponding row in the right table (22, 33, 44, 55, 66, ...)
+                out = pd.concat(
+                    [
+                        c1_df.iloc[leftIdx].reset_index(drop=True),   # select the rows of the left table
+                        c2_df.iloc[rightIdx].reset_index(drop=True)   # select the rows of the right table
+                    ], axis=1)  # concat the two tables "horizontally" (i.e., join columns, not append rows)
+
+                #save the order/pix/and distances for each nearest neighbor
+                out['hips_k'] = order
+                out['hips_pix'] = pix
+                out["_DIST"] =util.gc_dist(
+                    out[c1_md['ra_kw']], out[c1_md['dec_kw']],
+                    out[c2_md['ra_kw']], out[c2_md['dec_kw']]
+                )
+
+                #cull the return dataframe based on the distance threshold
+                out = out.loc[out['_DIST'] < dthresh]
+                retdfs.append(out)
+                #memory management
+                del out, dists, inds, outIdx, leftIdx, rightIdx, xy1, xy2
 
-        c1matches = [c1m for c1m in c1matches if c1m['_DIST'] <= dthresh]
-        if c1matches:
-            matches[id1] = c1matches
+            else:
+                retdfs.append(pd.DataFrame({}, columns=colnames))
+            del c1_df, c2_df
 
-    return len(matches)
+        except:
+            retdfs.append(pd.DataFrame({}, columns=colnames))
 
+    #if the npartitions are > 1 it will concatenate calculated catalogcrossmatches into a single dataframe to return
+    return pd.concat(retdfs)
 
 
 if __name__ == '__main__':
```

Comment:
necessary since the dask.dataframe.map_partitions(..) passes in a df that is iterable into the crossmatch func, so the return pandas.dataframe needs to be concatenated.